src/converter/one/convert/JfrToOtlp.java

```diff
@@ -0,0 +1,274 @@
+
+        // Write location table
+        for (Line l : linePool.keys()) {
+            long locMark = proto.startField(PROFILES_DICTIONARY_location_table, 1);
+            proto.field(LOCATION_mapping_index, 0);
+
+            long lineMark = proto.startField(LOCATION_line, 1);
+            proto.field(LINE_function_index, l.functionIdx);
+            proto.field(LINE_line, l.lineNumber);
+            proto.commitField(lineMark);
+
+            proto.commitField(locMark);
+        }
+
+        // Write string table
+        for (String s : stringPool.keys()) {
+            proto.field(PROFILES_DICTIONARY_string_table, s);
+        }
+
+        // Write attributes table
+        for (KeyValue kv : attributesPool.keys()) {
+            long aMark =
+                    proto.startField(PROFILES_DICTIONARY_attribute_table, BIG_MESSAGE_BYTE_COUNT);
+            proto.field(KEY_VALUE_key, kv.key);
+
+            long vMark = proto.startField(KEY_VALUE_value, BIG_MESSAGE_BYTE_COUNT);
+            proto.field(ANY_VALUE_string_value, kv.value);
+            proto.commitField(vMark);
+
+            proto.commitField(aMark);
+        }
+
+        proto.commitField(profilesDictionaryMark);
+    }
+
+    public static void convert(String input, String output, Arguments args) throws IOException {
+        JfrToOtlp converter;
+        try (JfrReader jfr = new JfrReader(input)) {
+            converter = new JfrToOtlp(jfr, args);
+            converter.convert();
+        }
+        try (FileOutputStream out = new FileOutputStream(output)) {
+            converter.dump(out);
+        }
+    }
+
+    private final class OtlpEventToSampleVisitor implements EventCollector.Visitor {
+        private final List<Integer> locationIndices;
+        private final double factor = counterFactor();
+
+        // JFR constant pool stacktrace ID to Range
+        private final Map<Integer, Range> idToRange = new HashMap<>();
+        // Next index to be used for a location into Profile.location_indices
+        private int nextLocationIdx = 0;
+
+        public OtlpEventToSampleVisitor(List<Integer> locationIndices) {
+            this.locationIndices = locationIndices;
+        }
+
+        @Override
+        public void visit(Event event, long samples, long value) {
+            long msFromStart = (event.time - jfr.chunkStartTicks) * 1_000 / jfr.ticksPerSec;
+            long timeNanos = jfr.chunkStartNanos + msFromStart * 1_000_000;
+
+            Range range = idToRange.computeIfAbsent(event.stackTraceId, this::computeLocationRange);
+
+            long sMark = proto.startField(PROFILE_sample, 1);
+            proto.field(SAMPLE_locations_start_index, range.start);
+            proto.field(SAMPLE_locations_length, range.length);
+            proto.field(SAMPLE_timestamps_unix_nano, timeNanos);
+
+            proto.field(
+                    SAMPLE_attribute_indices,
+                    attributesPool.index(new KeyValue("thread.name", getThreadName(event.tid))));
```

Comment:
My gut feeling tells me this can be a bottleneck due to lots of garbage created on the hot path.
I'm not asking you to rewrite it, though: it will be an interesting exercise to profile this later on.